### DISCUSSION TITLE
SOLVED Poor quality audio from IOS devices.

### DISCUSSION THREAD

**Discussion by mooballs** [pschatzmann/ESP32-A2DP#582] [Q&A]
I'd been having the same problem a few other people have with Android devices working more or less perfectly but IOS ones dropping out occasionally and stuttering and thought maybe it was a codec issue. I then bought a few of these [boards](https://www.aliexpress.com/item/1005006476900011.html?spm=a2g0o.order_list.order_list_main.47.73ed1802S4OJvn) and noticed that the problem varied considerably between boards (I'd tried changing I2S buffer, stack, etc to no avail) so thought it may be a hardware issue. After some searching I found [this](https://github.com/espressif/esp-idf/issues/11169) thread and thought I'd try changing the pin out for the external DAC (a PCM5102 like [this](https://www.onbuy.com/gb/p/24-bit-pcm5102-pcm5102a-stereo-dac-digital-to-analog-converter-pll-voice-module-phat~p89375975/?exta=gshp&extac=gshpfa) one). I changed the pinout from this:
`config.pin_bck = 26;
  config.pin_ws = 25;
  config.pin_data = 23;`
To this:
`config.pin_bck = 26;
  config.pin_ws = 25;
  config.pin_data = 19;` 
and although I haven't had time to go deep with it, it seems to work perfectly. In the original post it says maybe it's a muxing/pin-mapping issue, I wouldn't know and it won't work for everyone, but it's worth giving it a try.

**Comment by Rthunt15** [comment]
Do you know if any other pins would be "safe" as pin 19 is also MISO. I plan on using an SPI screen, so pin 19 wouldn't work in my case (an I'm sure many others)

**Reply by pschatzmann** [reply]
With the ESP32, basically all pins should work w/o issues for I2S.
You can also remap all SPI pins as well via the available API

If you are running into issues, it might be caused by cross-talk: so keep your wires short!

**Reply by Rthunt15** [reply]
Hey Phil, thanks for the quick response but that doesn't seem to be it. As the others in this thread is experiencing it seems to be isolated to IOS devices. Using my iPhone 13 the stuttering makes it unusable, but through my friends S24 and my PC the audio is perfect. Plus my wires are pretty sure, none of then longer than an inch. Good to know about SPI remapping though, didn't realize that was an option.

**Reply by pschatzmann** [reply]
With my very old IPAD it is also working w/o problems. 
You can can see in the open issues, it is only with newer Apple devices.

**Reply by Rthunt15** [reply]
Right, that's the problem that I'm trying to solve. Might just need to remap the SPI and call it a day rather than anything else though

**Reply by mooballs** [reply]
I would just breadboard it and try different pins according to what you have available, there's definitely variations between different boards.

**Comment by filopantilab** [comment]
Same issue here. Tried some pinmux variations with no luck: perfectly fine with any android device, skipping/stuttering with iphones.

**Comment by rhenry74** [comment]
what solves this?
